src/client/chat/systems.rs

```diff
@@ -0,0 +1,288 @@
+    for event in focus_events.read() {
+        match event.state {
+            FocusState::Unfocus => {
+                window.cursor.grab_mode = CursorGrabMode::Locked;
+                window.cursor.visible = false;
+            }
+            FocusState::Focus => {}
+        }
+    }
+}
+
+pub fn handle_chat_focus_player_events(
+    mut focus_change_events: EventReader<FocusChangeEvent>,
+    mut controller_query: Query<&mut FpsController>,
+) {
+    for event in focus_change_events.read() {
+        match event.state {
+            FocusState::Focus => {
+                for mut controller in &mut controller_query {
+                    controller.enable_input = false;
+                }
+            }
+            FocusState::Unfocus => {
+                for mut controller in &mut controller_query {
+                    controller.enable_input = true;
+                }
+            }
+        }
+    }
+}
+
+pub fn handle_chat_container_focus_events(
+    mut focus_change_events: EventReader<FocusChangeEvent>,
+    mut chat_container_query: Query<(
+        &mut BackgroundColor,
+        &mut chat_components::ChatMessageContainer,
+    )>,
+) {
+    let (mut background_color, mut chat_container) = chat_container_query.single_mut();
+    for event in focus_change_events.read() {
+        match event.state {
+            FocusState::Focus => {
+                background_color.0 = COLOR_FOCUSED;
+                chat_container.focused = true;
+            }
+            FocusState::Unfocus => {
+                background_color.0 = COLOR_UNFOCUSED;
+                chat_container.focused = false;
+            }
+        }
+    }
+}
+
+pub fn handle_chat_input_focus_events(
+    mut focus_change_events: EventReader<FocusChangeEvent>,
+    mut chat_input_query: Query<(
+        &mut BackgroundColor,
+        &mut chat_components::ChatMessageInputElement,
+    )>,
+) {
+    let (mut background_color, mut chat_container) = chat_input_query.single_mut();
+    for event in focus_change_events.read() {
+        match event.state {
+            FocusState::Focus => {
+                background_color.0 = COLOR_FOCUSED;
+                chat_container.focused = true;
+            }
+            FocusState::Unfocus => {
+                background_color.0 = COLOR_UNFOCUSED;
+                chat_container.focused = false;
+            }
+        }
+    }
+}
+
+pub fn handle_chat_input_system(
+    mut evr_kbd: EventReader<KeyboardInput>,
+    mut chat_input_query: Query<(&mut Text, &mut chat_components::ChatMessageInputElement)>,
+    mut event_reader: EventReader<chat_events::FocusChangeEvent>,
+    mut event_writer: EventWriter<chat_events::SendMessageEvent>,
+) {
+    if event_reader.read().len() != 0 {
+        return;
+    }
+
+    let (mut text, input_component) = chat_input_query.single_mut();
+
+    if !input_component.focused {
+        return;
+    }
+
+    let mut chat_input_value = match text.sections.first() {
```

Comment:
It was simpler to mutate this value than to handle different scopes with the borrow checker.